### PR TITLE
Represent tuples as arrays

### LINKED ISSUE #11: Represent tuples as as arrays
Tuples are currently represented as lists which has high memory overhead and makes attribute access slow.

### PR DESCRIPTION
**Pull request by bdahse**
Fixes #11 .

### CODE REVIEW

File: main/Fixpoint/Compiler.flix [hunk 1]
```diff
@@ -68,7 +68,8 @@ namespace Fixpoint {
             let augBody = augmentBody(body);
             let env = unifyVars(augBody);
             let join = joinTerms(env, augBody);
-            let ramTerms = List.map(compileHeadTerm(env), headTerms);
+            let hdTerms = List.toArray(headTerms) as & Pure;
```
Comment (bdahse) [inline main/Fixpoint/Compiler.flix:71]: These conversions are temporary until we migrate the AST to arrays.

**Comment by bdahse** [comment]
This adds 8-10 seconds to the transitive closure benchmark. I did't expect this change alone to provide any speedup on this benchmark but it shouldn't be slower?

**Comment by magnus-madsen** [comment]
A few comments:

- It is always good to get surprising results. Now you have something to write about and put into the report.
- This `pub opaque type Tuple[v] = Array[v]` boxes every array in an object (which despite the name `opaque` is not (yet) erased). This probably costs a lot of churn. So it would be better to use a type alias. (But can you, maybe not?)
- The benchmark only has binary relations, so the overhead is not so great. (Maybe design a new experiment where you have relations with 10 arguments and compare.

Next thing to try:

- Split Tuple into `Tuple1, Tuple2. TupleN, TupleA(Array)`. For a good choice of `N`. That is- specialize for the first N elements and then use an array when you go beyond that.

**Comment by bdahse** [comment]
> 
> 
> A few comments:
> 
>     * It is always good to get surprising results. Now you have something to write about and put into the report.
> 
>     * This `pub opaque type Tuple[v] = Array[v]` boxes every array in an object (which despite the name `opaque` is not (yet) erased). This probably costs a lot of churn. So it would be better to use a type alias. (But can you, maybe not?)
> 
The opaque type is to implement Order on Tuple. Would that work with a type alias? 
>     * The benchmark only has binary relations, so the overhead is not so great. (Maybe design a new experiment where you have relations with 10 arguments and compare.
> 
> 
> Next thing to try:
> 
>     * Split Tuple into `Tuple1, Tuple2. TupleN, TupleA(Array)`. For a good choice of `N`. That is- specialize for the first N elements and then use an array when you go beyond that.

**Comment by bdahse** [comment]
Okay you can implement Order on a type alias Array[v] if you move the instance to the topmost namespace. I suppose it now applies to all arrays though?
In any case, it didn't have a noticeable impact on performance. Arrays are still slower than lists on this benchmark.

**Comment by bdahse** [comment]
Does Flix have special syntax for projecting the i'th component of a tuple or would I have to match on them to extract e.g. the 4th component?

**Comment by magnus-madsen** [comment]
> 
> 
> Does Flix have special syntax for projecting the i'th component of a tuple or would I have to match on them to extract e.g. the 4th component?

Do you mean array indexing? `e[e]`?

**Comment by magnus-madsen** [comment]
Did you try JProfiler to see what kinds of objects are floating around?

**Comment by bdahse** [comment]
> 
> 
> > Does Flix have special syntax for projecting the i'th component of a tuple or would I have to match on them to extract e.g. the 4th component?
> 
> Do you mean array indexing? `e[e]`?

No I meant if I use actual Flix tuples. Does array indexing syntax work on them as well?

**Comment by bdahse** [comment]
> 
> 
> Did you try JProfiler to see what kinds of objects are floating around?

No not yet, but that may be the next step to get a better idea what is going on.

**Comment by magnus-madsen** [comment]
> > 
> > Do you mean array indexing? `e[e]`?
> 
> No I meant if I use actual Flix tuples. Does array indexing syntax work on them as well?

Ah- no, there is no way to index them without pattern matching.

**Comment by bdahse** [comment]
> 
> 
> > Did you try JProfiler to see what kinds of objects are floating around?
> 
> No not yet, but that may be the next step to get a better idea what is going on.

Running the profiler was rather insightful. Around 99% of the allocated objects are RedBlackTree nodes.
1.8 million nodes and another 1.8 million 5-tuple (I suppose these go with the nodes).
1.4m leaves. 1.4m Black color objects.
Also quite a few Comparison objects: 371k GreaterThan and 151k EqualTo and 6 LessThan (! why so few? Because the Edge relation looks like (0, 1), (1, 2), (2, 3) etc?) 
There are only 93k int arrays which I assume are my tuples.

So maybe it is not that there is no difference between arrays and lists. Memory (and time) is just completely dominated by the RedBlackTree. 

I have attached the [profile](https://github.com/flix/datalog-solver/files/6125096/tc.profile.zip) if you want to have a look at it yourself.

**Comment by magnus-madsen** [comment]
> 
> 
> > > Did you try JProfiler to see what kinds of objects are floating around?
> > 
> > 
> > No not yet, but that may be the next step to get a better idea what is going on.
> 
> Running the profiler was rather insightful. Around 99% of the allocated objects are RedBlackTree nodes.
> 1.8 million nodes and another 1.8 million 5-tuple (I suppose these go with the nodes).
> 1.4m leaves. 1.4m Black color objects.
> Also quite a few Comparison objects: 371k GreaterThan and 151k EqualTo and 6 LessThan (! why so few? Because the Edge relation looks like (0, 1), (1, 2), (2, 3) etc?)
> There are only 93k int arrays which I assume are my tuples.
> 
> So maybe it is not that there is no difference between arrays and lists. Memory (and time) is just completely dominated by the RedBlackTree.
> 
> I have attached the [profile](https://github.com/flix/datalog-solver/files/6125096/tc.profile.zip) if you want to have a look at it yourself.

Pretty useful information. Lets talk about it on monday. How many internal nodes does the RB tree have and does that match the expected number?

**Comment by magnus-madsen** [comment]
Could you paste a few screenshots?

**Comment by magnus-madsen** [comment]
This should make sure that there is only one instance of e.g. `Black` or of `Leaf`.

https://github.com/flix/flix/pull/1781

**Comment by magnus-madsen** [comment]
It would be nice, at the end of the computation (but before the result is discarded), to provide a snapshot of the non-GCed objects in the heap.

**Comment by bdahse** [comment]
[Screenshot of GCed objects](https://puu.sh/HosQ5/146417894d.png)
[Screenshot of live objects](https://puu.sh/HosR9/9f10c213b9.png)
[Screenshot of live objects](https://puu.sh/HosRB/391d61887d.png)
[Screenshot of GC activity](https://puu.sh/HosSj/222182c51a.png)

I'm not really sure what the difference between the two live object screenshots is. As in, how are the measurements different.

I don't understand your question about the number of internal nodes in the RB tree. I know the Path tree has 62500 keys and there should be one internal node per key. But it seems like the number of nodes we create in total (live and GCed) is excessive.
Every insert should allocate O(log n) nodes. We perform 62500 inserts on the Path relation so 62500 * log(62500) should be an upper bound on the nodes created (up to constant factors). But that number is less than a million and we allocate more than 17 times that. Also notice that the measurements only include 1/10 allocations (to keep the overhead manageable). So it's likely even worse than a factor 17 (not every missed allocation will be a node allocation so it's hard to say for sure how much larger, but it's probably not far off an additional factor 10 since the vast majority of allocations are node allocations).

We also insert into the intermediary NewPath tree, but that accounts for at most a factor 2.

**Comment by magnus-madsen** [comment]
(FYI: You can post images directly to GH)

**Comment by magnus-madsen** [comment]
Good observations: Let us talk about them on Monday.

As for why the numbers don't line up: I don't have a ready answer. Feel free to look into the RBTree. (It should be good, but maybe there is something). You can also look at the generated code. 

While there is a lot of churn for the colors, they don't actually take up any real space... so not sure how much https://github.com/flix/flix/pull/1781 would help.

**Comment by magnus-madsen** [comment]
OK, I guess I am wrong for GreaterEquals.

**Comment by bdahse** [comment]
Would it be possible to replace MutSet[Tuple[v]] with a TreeSet[Tuple[v]] from Java? Note that TreeSet is generic.

**Comment by magnus-madsen** [comment]
> 
> 
> Would it be possible to replace MutSet[Tuple[v]] with a TreeSet[Tuple[v]] from Java? Note that TreeSet is generic.

No, I think that would be very difficult.

**Comment by magnus-madsen** [comment]
Could the excessive allocations be from the `SearchEnv` and not from the database? I think that would explain it.

**Comment by bdahse** [comment]
> 
> 
> Could the excessive allocations be from the `SearchEnv` and not from the database? I think that would explain it.

Ah yes, we use RedBlackTrees all over the place. Good point. I will check that.

**Comment by bdahse** [comment]
Hmm it doesn't seem to come from SearchEnv. After [46c9c00](https://github.com/flix/datalog-solver/pull/13/commits/46c9c00f28bf1b7619c85c1928f2a83ae15d8807) there is no difference in running time and profiling shows it hardly changed the number of GCed nodes.
image
This allocation call graph shows that 99% of allocations happen in MutSet.addAll! (a wrapper around Set.union which ends up doing a bunch of RedBlackTree.insert):
image
We only call MutSet.addAll! one place: When we merge New into Full.

**Comment by magnus-madsen** [comment]
Union is an expensive operation. I am not sure how to estimate the amount of operations you would expect from it.

Do you think the new and full relations are correct? You don't have too many facts in new?

When merging two sets, it is best to merge the small into the large one. Maybe that could be done inside `MutSet`?

**Comment by magnus-madsen** [comment]
Does the newest nightly reduce the number of Red/Black/GreaterThan objects?

**Comment by magnus-madsen** [comment]
A guess: The union operation is always merging the larger tree into the smaller tree.

**Comment by bdahse** [comment]
> 
> 
> Union is an expensive operation. I am not sure how to estimate the amount of operations you would expect from it.
> 
> Do you think the new and full relations are correct? You don't have too many facts in new?
> 
The compiler emits a check that the tuple to be inserted into New is not in Full. It also purges New in every fixpoint iteration. I can doublecheck to make sure there is not a bug there. Otherwise New should generally be smaller than Full.
> When merging two sets, it is best to merge the small into the large one. Maybe that could be done inside `MutSet`?
>
>
It is possible that the union inside MutSet.addAll! inserts Full into New. That would be Bad(tm).

**Comment by bdahse** [comment]
> 
> 
> A guess: The union operation is always merging the larger tree into the smaller tree.

A look at Set.flix and MutSet.flix suggests you're right :)

**Comment by bdahse** [comment]
How do I compile my own "nightly Flix"?

**Comment by bdahse** [comment]
> 
> 
> Does the newest nightly reduce the number of Red/Black/GreaterThan objects?

It got rid of the comparions, GreaterThan, EqualTo etc. But it did nothing to Leaf and Color objects.

**Comment by magnus-madsen** [comment]
> 
> 
> How do I compile my own "nightly Flix"?

You wait or you pull the source and run gradle jar?

**Comment by bdahse** [comment]
> 
> 
> > How do I compile my own "nightly Flix"?
> 
> You wait or you pull the source and run gradle jar?

I figured it out. Apparently IntelliJ had not imported the gradle project.

**Comment by bdahse** [comment]
After swapping the order of arguments in MutSet.addAll! we are within 3-5 seconds of Flix. Node allocations have also dropped considerably. We are down to 2 million now!
image
In fact, RedBlackTree.foreach is now the allocation bottleneck, ahead of MutSet.addAll!
image
evalOp closures is 12.1% of that. Presumably those closures would go away if evalOp was tail recursive?

**Comment by magnus-madsen** [comment]
Great work!

Do you think Set.union could be fixed in Flix? E.g. by computing the sizes (or perhaps just heights) of the trees before the union?

**Comment by magnus-madsen** [comment]
By the way -- all these small improvements -- collect at list of them for the report. It shows breadth.

**Comment by magnus-madsen** [comment]
>  evalOp closures is 12.1% of that. Presumably those closures would go away if evalOp was tail recursive?

Possibly, but probably also because of `|>`.

Using let-bound variables instead of `|>` might decrease the number of closure allocations.

**Comment by magnus-madsen** [comment]
> 
> 
> > evalOp closures is 12.1% of that. Presumably those closures would go away if evalOp was tail recursive?
> 
> Possibly, but probably also because of `|>`.
> 
> Using let-bound variables instead of `|>` might decrease the number of closure allocations.

The interpreter is less than 150 lines, so I think it is okay to write it in a slightly less elegant style.

The RAM compiler and all other infrastructure can still use` |>` and other convenience functions.

**Comment by magnus-madsen** [comment]
Let me know when you want to merge this PR?

**Comment by bdahse** [comment]
> 
> 
> Great work!
> 
> Do you think Set.union could be fixed in Flix? E.g. by computing the sizes (or perhaps just heights) of the trees before the union?

Yes, Set.union is the right place to fix it imo. Computing sizes takes linear time so perhaps a bit expensive. Height can probably do it, as you say. RB Trees are balanced to within a constant factor so height should be roughly equivalent to size.

I don't think the (advertised) time complexity of Set.union is quite right. It's claimed to be ` O(Set.size(ys) * Int32.log2(Set.size(xs)))` but the implementation looks more like ` O(Set.size(ys) * Int32.log2(Set.size(xs) + Set.size(ys)))` to me. The difference between the two can be meaningful when `Set.size(ys) >> Set.size(xs)`. 

With the new implementation it should be `O(Int32.min(Set.size(xs), Set.size(ys)) * Int32.log2(Int32.max(Set.size(xs), Set.size(ys)))`

**Comment by bdahse** [comment]
> 
> 
> Let me know when you want to merge this PR?

I think this PR has served its purpose now.

**Comment by bdahse** [comment]
Actually why use height instead of size? Both take linear time to compute.

**Comment by magnus-madsen** [comment]
> 
> 
> Actually why use height instead of size? Both take linear time to compute.

Why? You can compute the height in log n time.

(Actually you should compute the black height and compare them!)

**Comment by magnus-madsen** [comment]
Will you make a PR for an optimized version of Set.union to the Flix repo?

**Comment by bdahse** [comment]
> 
> 
> > Actually why use height instead of size? Both take linear time to compute.
> 
> Why? You can compute the height in log n time.
> 
> (Actually you should compute the black height and compare them!)

Ah I see what you mean. The black height of the root is the same for all paths so we can just go left. 

> 
> 
> Will you make a PR for an optimized version of Set.union to the Flix repo?

Yes.

**Comment by magnus-madsen** [comment]
Maybe you could open an issue to track the performance over time? and after specific changes?

**Pull request merged**